scripts/packages_exporter.py

```diff
@@ -348,59 +377,62 @@ async def check_noarch_in_user_distribution_repos(self, distr_name: str):
+                        for key_id in key_ids_lower:
+                            sub_keys = self.known_subkeys.get(key_id, [])
+                            if sig in sub_keys:
+                                return SignStatusEnum.SUCCESS, sig
+
+                return SignStatusEnum.WRONG_SIGNATURE, sig
+
         errored_packages = set()
         no_signature_packages = set()
         wrong_signature_packages = set()
-        for package in os.listdir(repository_path):
-            package_path = os.path.join(repository_path, package)
-            if not package_path.endswith('.rpm'):
-                self.logger.debug('Skipping non-RPM file or directory: %s',
-                                  package_path)
-                continue
-            args = ('rpm', '-qip', package_path)
-            exit_code, out, err = local['sudo'].run(args=args, retcode=None)
-            if exit_code != 0:
-                self.logger.error('Cannot get information about package %s, %s',
-                                  package_path, '\n'.join((out, err)))
-                errored_packages.add(package_path)
-                continue
-            signature_line = None
-            for line in out.split('\n'):
-                line = line.strip()
-                if line.startswith('Signature'):
-                    signature_line = line
-                    break
-            if not signature_line:
-                self.logger.error('No information about package %s signature',
-                                  package_path)
-                continue
-            signature_result = signature_regex.search(signature_line)
-            if not signature_result:
-                self.logger.error('Cannot detect information '
-                                  'about package %s signature', package_path)
-                errored_packages.add(package_path)
-                continue
-            pkg_key_id = signature_result.groupdict().get('key_id', '').lower()
-            if 'none' in pkg_key_id:
-                self.logger.error('Package %s is not signed', package_path)
-                no_signature_packages.add(package_path)
-            elif pkg_key_id not in key_ids_lower:
-                # Check if package is signed with known sub-key
-                signed_with_subkey = False
-                for key, subkeys in self.known_subkeys.items():
-                    if pkg_key_id in subkeys:
-                        signed_with_subkey = True
-                        break
-                if not signed_with_subkey:
-                    self.logger.error('Package %s is signed with wrong key, '
-                                      'expected "%s", got "%s"',
-                                      package_path, str(key_ids_lower),
-                                      pkg_key_id)
-                    wrong_signature_packages.add(f'{package_path} {pkg_key_id}')
+        futures = {}
+
+        with ThreadPoolExecutor(max_workers=10) as executor:
```

Comment:
can we use here `asyncio.gather`?
i may be wrong, but it's kinda strange or maybe wrong to call `ThreadPoolExecutor` inside `async` func that have eventloop and etc.

Reply:
The function for checking signatures is synchronous, RPM bindings are also synchronous, I've tested this approach and everything is working fine with this setup. Calling synchronous code in async function is not prohibited, it's just locking the context until function is completed